File: server/src/main/java/org/codi/catan/impl/game/SetupMoveHelper.java
/*
 * @author the-codinator
 * created on 2020/5/30
 */

package org.codi.catan.impl.game;

import javax.inject.Inject;
import javax.inject.Singleton;
import org.codi.catan.core.CatanException;
import org.codi.catan.model.game.Board;
import org.codi.catan.model.game.Color;
import org.codi.catan.model.game.House;
import org.codi.catan.model.game.HouseType;
import org.codi.catan.model.game.Phase;
import org.codi.catan.model.game.Road;
import org.codi.catan.model.game.State;
import org.codi.catan.model.request.SetupMoveRequest;

@Singleton
public class SetupMoveHelper {

    private final GraphHelper graphHelper;
    private final GameUtility gameUtility;
    private final BuildMoveHelper buildMoveHelper;
    private final MiscMoveHelper miscMoveHelper;

    @Inject
    public SetupMoveHelper(GraphHelper graphHelper, GameUtility gameUtility, BuildMoveHelper buildMoveHelper,
        MiscMoveHelper miscMoveHelper) {
        this.graphHelper = graphHelper;
        this.gameUtility = gameUtility;
        this.buildMoveHelper = buildMoveHelper;
        this.miscMoveHelper = miscMoveHelper;
    }

    /**
     * Play setup 1 or 2
     */
    public void play(Board board, State state, SetupMoveRequest request) throws CatanException {
        // Validate input
        graphHelper.validateVertex(request.getHouseVertex());
        graphHelper.validateVertex(request.getRoadVertex());
        buildMoveHelper.ensureCanPlaceHouse(state, request.getHouseVertex(), HouseType.settlement);
        buildMoveHelper.ensureCanPlaceRoad(state, request.getHouseVertex(), request.getRoadVertex());
        // Create settlement & road
        Color color = state.getCurrentMove().getColor();
        state.getHouses().put(request.getHouseVertex(), new House(color, HouseType.settlement));
        state.getRoads().add(new Road(color, request.getHouseVertex(), request.getRoadVertex()));
        // Gain resources
        if (state.getPhase() == Phase.setup2) {
            for (int hex : graphHelper.getConnectedHexListForVertex(request.getHouseVertex())) {
                gameUtility.transferResources(state, null, color, board.getTiles()[hex].getResource());
            }
        }
        // Auto end turn
        miscMoveHelper.endTurn(board, state);
    }
}

File: server-azure/src/impl/game/setup-move-helper.ts
import { ensureCanPlaceHouse, ensureCanPlaceRoad } from './build-move-helper';
import { getConnectedHexListForVertex, validateVertex } from './graph-helper';
import { HouseType } from '../../model/game/house-type';
import { Phase } from '../../model/game/phase';
import type { PlayOptions } from './move-api-helper';
import type { SetupMoveRequest } from '../../model/request/game-request';
import { endTurn } from './misc-move-helper';
import { transferResourcesList } from './game-utility';

export function play({ board, state, request }: PlayOptions<SetupMoveRequest>): void {
  validateVertex(request.houseVertex);
  validateVertex(request.roadVertex);

  ensureCanPlaceHouse(state, request.houseVertex, HouseType.settlement);
  ensureCanPlaceRoad(state, request.houseVertex, request.roadVertex);
  // Create settlement & road
  const color = state.currentMove.color;
  state.houses[request.houseVertex] = { color, type: HouseType.settlement };
  state.roads.push({ color, vertex1: request.houseVertex, vertex2: request.roadVertex });
  // Gain resources
  if (state.phase === Phase.setup2) {
    for (const hex of getConnectedHexListForVertex(request.houseVertex)) {
      transferResourcesList(state, undefined, color, board.tiles[hex].resource);
    }
  }
  // Auto end turn
  endTurn(board, state);
}
